### PR TITLE
Mechanical Changes to ProvisionerTask

### PR DESCRIPTION
**Pull request by manadart**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

Mechanical changes to `ProvisionerTask` including:
- Some modularisation.
- Error checking and tracing.
- Import name conflicts.
- Line lengths.

## QA steps

No functional changes. Tests remain green.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[ProvisionerTask Mechanics](https://trello.com/c/hUlGWaHt/2292-provisionertask-mechanics)

File: worker/provisioner/provisioner_task.go [hunk 1]
```diff
@@ -506,27 +461,88 @@ func (task *provisionerTask) findUnknownInstances(stopping []instances.Instance)
...
+	// Remove any dead machines from state.
+	for _, machine := range dead {
+		task.logger.Infof("removing dead machine %q", machine.Id())
+		if err := machine.MarkForRemoval(); err != nil {
+			task.logger.Errorf("failed to remove dead machine %q", machine.Id())
+		}
+		task.removeMachineFromAZMap(machine)
+		task.machinesMutex.Lock()
+		delete(task.machines, machine.Id())
+		task.machinesMutex.Unlock()
```
Comment (SimonRichardson) [inline worker/provisioner/provisioner_task.go]: I know this is a mechanical change, but I'd encapsulate this in another method to improve readability.

File: worker/provisioner/provisioner_task.go [hunk 2]
```diff
@@ -1334,8 +1351,10 @@ func volumeAttachmentsToAPIServer(attachments []storage.VolumeAttachment) map[st
 	for _, a := range attachments {
 		var planInfo *params.VolumeAttachmentPlanInfo
 		if a.PlanInfo != nil {
-			planInfo.DeviceType = a.PlanInfo.DeviceType
-			planInfo.DeviceAttributes = a.PlanInfo.DeviceAttributes
+			planInfo = &params.VolumeAttachmentPlanInfo{
+				DeviceType:       a.PlanInfo.DeviceType,
+				DeviceAttributes: a.PlanInfo.DeviceAttributes,
+			}
```
Comment (SimonRichardson) [inline worker/provisioner/provisioner_task.go]: Hold on a minute, this actually fixes a runtime panic surely?

File: worker/provisioner/provisioner_task.go [hunk 3]
```diff
@@ -1236,17 +1233,25 @@ func (task *provisionerTask) startMachine(
...
+		return machineProfiles, nil
+	}
+
+	manager, ok := task.broker.(container.LXDProfileNameRetriever)
+	if !ok {
+		task.logger.Tracef("failed to gather profile names; broker does not implement LXDProfileNameRetriever")
+		return machineProfiles, nil
+	}
+
+	profileNames, err := manager.LXDProfileNames(instanceId)
```
Comment (SimonRichardson) [inline worker/provisioner/provisioner_task.go]: I can't remember why we didn't use this error, it seems like an oversight!?

:+1:

File: worker/provisioner/provisioner_task.go (hunk 2 above)
Comment (manadart) [reply worker/provisioner/provisioner_task.go]: Volume attachment plans (I think) are only used to attach SCSI volumes to OCI instances.

It is possible that no one has added a machine with a volume attachment specified at creation time. I think if one were to attach such a volume to an existing machine, this path would not be traversed.

I could do some A/B testing, but the "popularity" of that provider doesn't justify verifying that this fixes a specific error. I'm happy that it is correct *now*.

Comment (SimonRichardson) [reply worker/provisioner/provisioner_task.go]: My worry is that we don't know how this should work, I'm assuming it should work like this, but we're unsure. Can we at least leave a note stating that fact?

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**